lib/puppet/agent.rb

```diff
@@ -60,6 +61,23 @@ def run(client_options = {})
               end
             end
           rescue Puppet::LockError
+            if client_options[:waitforlock]
+              wait_for_lock_deadline ||= Time.now.to_i + Puppet[:maxwaitforlock]
```

Comment:
Since `wait_for_lock_deadline` is local, won't we reset this each time a `LockError` is raised, so the deadline will always be `maxwaitforlock` seconds from now? I'd kind of expect `wait_for_lock_deadline = nil` on line 45, do that it's only set once (the first time the LockError is raised).